### ISSUE TITLE
Connection refused on Drobo FS

### ISSUE THREAD

**Issue by spl** [droboports/openssh#1]
Sorry for the vague issue, but I'm not sure why OpenSSH didn't work. This is the first DroboApp I'm attempting to install, so it could be something I did.

I copied the released 6.7 openssh.tgz to DroboApps and restarted, but I couldn't connect.

In the log, I saw:

```
Fri Dec 26 04:59:51 2014: DNASConfig::DroboAppsStartStop: INFO: /usr/bin/DroboApps.sh returned success. Output:
Fri Dec 26 04:59:51 2014: =========================================================
Fri Dec 26 04:59:51 2014: Installing /mnt/DroboFS/Shares/DroboApps/openssh.tgz.
Running post-install script.
Starting openssh.

Fri Dec 26 04:59:51 2014: =========================================================
Fri Dec 26 04:59:51 2014: DNASConfigInitThread: LoadConfigAndStart Finished correctly
```

and

```
Fri Dec 26 05:22:46 2014: DNASConfig::DroboAppsStartStop: INFO: /usr/bin/DroboApps.sh returned success. Output:
Fri Dec 26 05:22:46 2014: =========================================================
Fri Dec 26 05:22:46 2014: Starting openssh.

Fri Dec 26 05:22:46 2014: =========================================================
Fri Dec 26 05:22:46 2014: DNASConfigInitThread: LoadConfigAndStart Finished correctly
```

I tried disabling and re-enabling DroboApps. I restarted several times. Nothing changed anything.

I thought to try the Dropbear SSH app, since that is supposedly already installed. It seems like Data Robotics is no longer providing the `.tgz`, but your old site still has a link [here](https://sites.google.com/a/droboports.com/www/using-command-line-apps). Fortunately, SSH now works. This leaves me to believe there may be a problem with the OpenSSH app.

**Comment by ricardopadilha** [comment]
Do you have both Dropbear and OpenSSH open at the same time? If so, please keep in mind that they will fight over port 22. If you can, please have a look at `/tmp/DroboApps/openssh/log.txt`. That file will contain a detailed log of the launch of OpenSSH. It is possible that the file won't be there, since `/tmp` is reset every boot. We need to have a look at that log file because as far as I can tell from what you are showing here, everything seems to have installed fine.

**Comment by spl** [comment]
> Do you have both Dropbear and OpenSSH open at the same time?

No, OpenSSH was the first DroboApp I tried to install. When that didn't work, I deleted `DroboApps/openssh`, restarted, and installed `DroboApps/dropbear`.

> If you can, please have a look at `/tmp/DroboApps/openssh/log.txt`.

```
# ls -l /tmp/
-rw-r--r--    1 root     root            0 Jan 20  2009 netperf.debug
```

> We need to have a look at that log file because as far as I can tell from what you are showing here, everything seems to have installed fine.

How do you suggest I get that log file? If I uninstall Dropbear and install OpenSSH and I can't ssh into the box, then I'm not sure how to get the file.

**Comment by ricardopadilha** [comment]
You can edit the service.sh of either Dropbear of OpenSSH to change the port. For dropbear, for example, you can just add `-p 2222` to its command line to change its port to 2222.

**Comment by spl** [comment]
Good idea. I'm busy at the moment, but I'll try it later, maybe over the weekend.

**Comment by spl** [comment]
Okay. I installed OpenSSH. Here are the logs from startup:

```
# pwd
/tmp/DroboApps/openssh
# ls -l
-rw-r--r--    1 root     root          817 Dec 27 08:45 install.log
-rw-r--r--    1 root     root          464 Dec 27 08:45 log.txt
# cat install.log
2014-12-27 08-44-35: /mnt/DroboFS/Shares/DroboApps/openssh/install.sh
+ basename /mnt/DroboFS/Shares/DroboApps/openssh/etc/ssh_config.default .default
+ basefile=/mnt/DroboFS/Shares/DroboApps/openssh/etc/ssh_config
+ [ ! -f /mnt/DroboFS/Shares/DroboApps/openssh/etc/ssh_config ]
+ cp /mnt/DroboFS/Shares/DroboApps/openssh/etc/ssh_config.default /mnt/DroboFS/Shares/DroboApps/openssh/etc/ssh_config
+ basename /mnt/DroboFS/Shares/DroboApps/openssh/etc/sshd_config.default .default
+ basefile=/mnt/DroboFS/Shares/DroboApps/openssh/etc/sshd_config
+ [ ! -f /mnt/DroboFS/Shares/DroboApps/openssh/etc/sshd_config ]
+ cp /mnt/DroboFS/Shares/DroboApps/openssh/etc/sshd_config.default /mnt/DroboFS/Shares/DroboApps/openssh/etc/sshd_config
+ bin/ssh-keygen -A
ssh-keygen: generating new host keys: RSA1 RSA DSA ECDSA ED25519
# cat log.txt
2014-12-27 08-45-11: /mnt/DroboFS/Shares/DroboApps/openssh/service.sh start
+ _service_start
+ _is_running /tmp/DroboApps/openssh/pid.txt
+ /sbin/start-stop-daemon -K -s 0 -x /mnt/DroboFS/Shares/DroboApps/openssh/sbin/sshd -p /tmp/DroboApps/openssh/pid.txt -q
+ set +x
+ _create_user
+ id -u sshd
id: unknown user sshd
+ /mnt/DroboFS/Shares/DroboApps/openssh/libexec/useradd -r -M -d /mnt/DroboFS/Shares/DroboApps/openssh/var/empty -s /bin/false -g 99 -u 103 sshd
```

I didn't see `sshd` in `ps`. Here's the log when I tried to run it manually:

```
2014-12-27 08-50-25: ./service.sh restart
+ _service_restart
+ _service_stop
+ /sbin/start-stop-daemon -K -x /mnt/DroboFS/Shares/DroboApps/openssh/sbin/sshd -p /tmp/DroboApps/openssh/pid.txt -v
no /mnt/DroboFS/Shares/DroboApps/openssh/sbin/sshd found; none killed
+ echo openssh is not running
+ sleep 3
+ _service_start
+ _is_running /tmp/DroboApps/openssh/pid.txt
+ /sbin/start-stop-daemon -K -s 0 -x /mnt/DroboFS/Shares/DroboApps/openssh/sbin/sshd -p /tmp/DroboApps/openssh/pid.txt -q
+ set +x
+ _create_user
+ id -u sshd
id: unknown user sshd
+ /mnt/DroboFS/Shares/DroboApps/openssh/libexec/useradd -r -M -d /mnt/DroboFS/Shares/DroboApps/openssh/var/empty -s /bin/false -g 99 -u 103 sshd
```

**Comment by ricardopadilha** [comment]
Everything seems to be ok up to the creation of the sshd user. Can you please try this command directly on the prompt, and tell me if it works? 

`/mnt/DroboFS/Shares/DroboApps/openssh/libexec/useradd -r -M -d /mnt/DroboFS/Shares/DroboApps/openssh/var/empty -s /bin/false -g 99 -u 103 sshd`

I tried on my FS and it seems to work. I just noticed that there is a problem with the generation of the PID file on the FS, which makes the app appear as "not started" on the dashboard but that does not prevent OpenSSH from actually running (i.e., you can see it in `top` and `ps`).

**Comment by spl** [comment]
> Can you please try this command directly on the prompt, and tell me if it works?

I ran the command exactly as stated, but I cannot find any signs that it worked. `/etc/passwd`, `/etc/.passwd`, `/etc/group`, and `/etc/shadow` are unchanged. `/etc/passwd.lock` and `/etc/group.lock` were updated, presumably with the process ID of `useradd`.

> I just noticed that there is a problem with the generation of the PID file on the FS, which makes the app appear as "not started" on the dashboard but that does not prevent OpenSSH from actually running.

Could you please explain how you see an app on the dashboard? Are you referring to the Drobo Dashboard? I don't see anything there about apps except for the Admin dialog, which has the checkbox for enabling DroboApps.

**Comment by ricardopadilha** [comment]
Is by any chance `/` full? That is the only explanation I can think of for the `passwd` not being updated. There should be a line starting with `sshd` in there.

The Dashboard I mentioned is the Drobo Dashboard, yes. But if you have an FS there won't be support for DroboApps in there.

**Comment by ProbablePattern** [comment]
Same issue with connection refused on a Drobo FS.

**Comment by spl** [comment]
@ricardopadilha:

> Is by any chance `/` full?

```
# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                21.0M     19.0M      2.0M  91% /
...
```

That's not too full, is it?

> But if you have an FS there won't be support for DroboApps in there.

Ah, okay. I just realized I didn't explicitly say I was installing it on an FS, which I am. I updated the issue title to be clear.

**Comment by spl** [comment]
I was able to get OpenSSH running with the following steps.
1. Delete all lock files that might be left over from running `useradd`:
   
   ```
   # cd /etc
   # rm -f passwd.lock group.lock .pwd.lock
   ```
   
   I did this as a precaution. I'm not sure whether or not the next step uses the lock files.
2. Use the included ([BusyBox](http://www.busybox.net/downloads/BusyBox.html)) `adduser` command to create the user `sshd`:
   
   ```
   # /bin/adduser -H -S -s /bin/false -h /mnt/DroboFS/Shares/DroboApps/openssh/var/empty sshd
   ```
   
   To check if the user was added, do the following:
   
   ```
   # id sshd
   uid=1(sshd) gid=1(sshd) groups=1(sshd)
   ```
   
   Note that these are the default UID, GID, and groups. I didn't change them. In fact, I couldn't figure out how to change the UID and GID to the ones used by the OpenSSH `service.sh` script without directly editing `/etc/passwd` and `/etc/group`. But it turned out that this was not necessary.
3. Run the `openssh` service:
   
   ```
   # /mnt/DroboFS/Shares/DroboApps/openssh/service.sh start
   ```
   
   I confirmed that `sshd` was running by looking at `ps` as well as `ssh`ing in.

This indicates to me that the problem lies with the `useradd` command included with the OpenSSH distribution.

**Comment by ricardopadilha** [comment]
Interesting. One last question: do you have a `/etc/login.defs` file? The only reason I can think of for useradd not working is that this file somehow did not exist. `service.sh` should have created it, but if it did not can you please do `touch /etc/login.defs` and try the useradd command again?

**Comment by spl** [comment]
Well, that's interesting. I just turned my DroboFS on for the first time since my last comment. Now, OpenSSH is starting up properly:

```
# cat /tmp/DroboApps/openssh/log.txt
2015-01-01 00-01-00: /mnt/DroboFS/Shares/DroboApps/openssh/service.sh start
+ _service_start
+ _is_running /tmp/DroboApps/openssh/pid.txt
+ /sbin/start-stop-daemon -K -s 0 -x /mnt/DroboFS/Shares/DroboApps/openssh/sbin/sshd -p /tmp/DroboApps/openssh/pid.txt -q
+ set +x
+ _create_user
+ id -u sshd
id: unknown user sshd
+ /mnt/DroboFS/Shares/DroboApps/openssh/libexec/useradd -r -M -d /mnt/DroboFS/Shares/DroboApps/openssh/var/empty -s /bin/false -g 99 -u 103 sshd
+ _fix_permissions
+ local devices=/dev/null /dev/full /dev/random /dev/urandom /dev/tty /dev/ptmx /dev/zero /dev/crypto
+ [[ -c /dev/null ]]
+ stat -c %a /dev/null
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/null
+ [[ -c /dev/full ]]
+ stat -c %a /dev/full
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/full
+ [[ -c /dev/random ]]
+ stat -c %a /dev/random
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/random
+ [[ -c /dev/urandom ]]
+ stat -c %a /dev/urandom
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/urandom
+ [[ -c /dev/tty ]]
+ stat -c %a /dev/tty
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/tty
+ [[ -c /dev/ptmx ]]
+ stat -c %a /dev/ptmx
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/ptmx
+ [[ -c /dev/zero ]]
+ stat -c %a /dev/zero
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/zero
+ [[ -c /dev/crypto ]]
+ stat -c %a /dev/crypto
+ [[ 660 -ne 666 ]]
+ chmod a+rw /dev/crypto
+ [[ ! -f /var/log/lastlog ]]
+ [[ ! -f /etc/login.defs ]]
+ chmod 4711 /mnt/DroboFS/Shares/DroboApps/openssh/libexec/ssh-keysign
+ /mnt/DroboFS/Shares/DroboApps/openssh/sbin/sshd
```

Indeed, `/etc/login.defs` is there now. By the file's modification date, it appears that it was updated when I was previously working on this. I deleted `/etc/login.defs` and the `useradd` command failed just as before. I `touch`ed it, and it worked.

I tried changing the order of operations in `service.sh` so that `_fix_permissions` happened before `_create_user` (after `rm /etc/login.defs`):

```
--- service.sh- Thu Nov 27 04:27:11 2014
+++ service.sh  Thu Jan  1 08:43:04 2015
@@ -50,8 +50,8 @@
   set -u # exit on unset variable
   set -e # exit on uncaught error code
   set -x # enable script trace
-  _create_user
   _fix_permissions
+  _create_user
   "${daemon}"
 }
```

That allowed OpenSSH to start up but, interestingly, Dropbear failed to start. If I just put the `/etc/login.defs` part first, then both OpenSSH and Dropbear started up:

```
--- service.sh- Thu Nov 27 04:27:11 2014
+++ service.sh  Thu Jan  1 09:03:16 2015
@@ -50,6 +50,7 @@
   set -u # exit on unset variable
   set -e # exit on uncaught error code
   set -x # enable script trace
+  if [[ ! -f /etc/login.defs ]]; then touch /etc/login.defs; fi
   _create_user
   _fix_permissions
   "${daemon}"
```

Maybe that's a change you should consider?

**Comment by ricardopadilha** [comment]
Absolutely. This is actually a regression, the old version had this line exactly as you indicated. I probably lost it during the migration to github.

**Comment by spl** [comment]
Cool. I'm just a bit concerned about the issue I had where Dropbear didn't start. But I tried your change to `service.sh`, and, after `rm /etc/login.defs` and rebooting, both Dropbear and OpenSSH started.

**Comment by spl** [comment]
Lastly, do you want to need a new release? The [latest release](https://github.com/droboports/openssh/releases/tag/v6.7-fs) says "Includes fix for issue #1" but also has "[1 commit](https://github.com/droboports/openssh/releases/tag/v6.7-fs) to drobofs since this release", which, I believe, means 00361e9 is not included in the release tag.

**Comment by ricardopadilha** [comment]
Sorry about that. I update the release tag and everything should be ok now.
